node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -207,14 +235,94 @@ impl OrderingProvider {
 	}
 }
 
+async fn send_message_fatal<Sender, Response>(
+	sender: &mut Sender,
+	message: impl Into<AllMessages>,
+	receiver: oneshot::Receiver<std::result::Result<Response, ChainApiError>>,
+) -> FatalResult<Response>
+where
+	Sender: SubsystemSender,
+{
+	sender.send_message(message.into()).await;
+
+	receiver
+		.await
+		.map_err(|_| Fatal::ChainApiSenderDropped)?
+		.map_err(Fatal::ChainApi)
+}
+
 async fn get_block_number(
 	sender: &mut impl SubsystemSender,
 	relay_parent: Hash,
 ) -> FatalResult<Option<BlockNumber>> {
 	let (tx, rx) = oneshot::channel();
-	sender.send_message(ChainApiMessage::BlockNumber(relay_parent, tx).into()).await;
+	send_message_fatal(sender, ChainApiMessage::BlockNumber(relay_parent, tx), rx).await
+}
+
+async fn get_finalized_block_hash(sender: &mut impl SubsystemSender) -> FatalResult<Option<Hash>> {
+	let (number_tx, number_rx) = oneshot::channel();
+	let number =
+		send_message_fatal(sender, ChainApiMessage::FinalizedBlockNumber(number_tx), number_rx)
+			.await?;
+
+	let (hash_tx, hash_rx) = oneshot::channel();
+	send_message_fatal(sender, ChainApiMessage::FinalizedBlockHash(number, hash_tx), hash_rx).await
+}
+
+/// Returns ancestor hashes of `head` in the descending order, stopping
+/// either at `last_observed_hash` or the genesis block.
+///
+/// Suited specifically for querying non-finalized chains, thus doesn't
+/// rely on block numbers.
+///
+/// The caller must ensure that `last_observed_hash` is indeed an ancestor
+/// of `head` in order to avoid traversing the whole chain.
+/// Both `head` and `last_observed_hash` are **not** included in the result.
+async fn get_block_ancestors_hashes<Sender>(
+	sender: &mut Sender,
+	mut head: Hash,
+	last_observed_hash: Hash,
+) -> NonFatalResult<Vec<Hash>>
+where
+	Sender: SubsystemSender,
+{
+	const ANCESTRY_STEP: usize = 4;
```

Comment:
Might be reasonable to increase this value during start up in case the node is in major sync

Reply:
Assuming finalization is not stalled, even on startup this will only be a couple of blocks. So I think, we don't need to worry about that. Maybe increase it to 10 in general or something.